File: artifacts/haskell/src/MySpace/MusicCatalogue/Statements/InsertAlbum.hs
module MySpace.MusicCatalogue.Statements.InsertAlbum where

import MySpace.MusicCatalogue.Prelude
import Test.QuickCheck
import qualified Hasql.Statement
import qualified Hasql.Decoders
import qualified Hasql.Encoders
import qualified Data.Aeson
import qualified Data.Vector
import qualified Hasql.Mapping.IsStatement
import qualified Hasql.Mapping.IsScalar
import MySpace.MusicCatalogue.Types
import qualified PostgresqlTypes
import qualified PostgresqlTypes.Date
import qualified PostgresqlTypes.Bytea
import qualified PostgresqlTypes.Numeric
import qualified PostgresqlTypes.Float4
import qualified PostgresqlTypes.Float8
import qualified PostgresqlTypes.Int2
import qualified PostgresqlTypes.Int4
import qualified PostgresqlTypes.Int8
import qualified PostgresqlTypes.Text
import qualified PostgresqlTypes.Uuid

-- |
-- Parameters for the @insert_album@ query.
--
-- ==== SQL Template
--
-- > insert into album (name, released, format, recording)
-- > values ($name, $released, $format, $recording)
-- > returning id
--
-- ==== Source Path
--
-- > ./queries/insert_album.sql
--
data InsertAlbum = InsertAlbum
  { -- | Maps to @name@.
    name :: Text,
    -- | Maps to @released@.
    released :: PostgresqlTypes.Date,
    -- | Maps to @format@.
    format :: AlbumFormat,
    -- | Maps to @recording@.
    recording :: RecordingInfo
  }
  deriving stock (Eq, Show)

-- | Result of the statement parameterised by 'InsertAlbum'.
type InsertAlbumResult = InsertAlbumResultRow

-- | Row of 'InsertAlbumResult'.
newtype InsertAlbumResultRow = InsertAlbumResultRow
  { -- | Maps to @id@.
    id :: Int64
  }
  deriving stock (Show, Eq)

instance Arbitrary InsertAlbum where
  arbitrary =
    InsertAlbum
      <$> (PostgresqlTypes.Text.toText <$> arbitrary)
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
      
instance Hasql.Mapping.IsStatement.IsStatement InsertAlbum where
  type Result InsertAlbum = InsertAlbumResult

  statement = Hasql.Statement.preparable sql encoder decoder
    where
      sql =
        "insert into album (name, released, format, recording)\n\
        \values ($1, $2, $3, $4)\n\
        \returning id"

      encoder =
        mconcat
          [ (.name) >$< Hasql.Encoders.param (Hasql.Encoders.nonNullable (Hasql.Mapping.IsScalar.encoder)),
            (.released) >$< Hasql.Encoders.param (Hasql.Encoders.nonNullable (Hasql.Mapping.IsScalar.encoder)),
            (.format) >$< Hasql.Encoders.param (Hasql.Encoders.nonNullable (Hasql.Mapping.IsScalar.encoder)),
            (.recording) >$< Hasql.Encoders.param (Hasql.Encoders.nonNullable (Hasql.Mapping.IsScalar.encoder))
          ]

      decoder =
        Hasql.Decoders.singleRow do
          id <- Hasql.Decoders.column (Hasql.Decoders.nonNullable (Hasql.Mapping.IsScalar.decoder))
          pure InsertAlbumResultRow {..}

File: artifacts/rust/src/statements/insert_album.rs
use tokio_postgres::types::Type;

/// Parameters for the `insert_album` query.
///
/// # SQL Template
///
/// ```sql
/// insert into album (name, released, format, recording)
/// values ($name, $released, $format, $recording)
/// returning id
/// ```
///
/// # Source Path
///
/// `./queries/insert_album.sql`
#[derive(Debug, Clone, PartialEq, Default)]
pub struct Input {
    /// Maps to `$name` in the template.
    pub name: String,
    /// Maps to `$released` in the template.
    pub released: chrono::NaiveDate,
    /// Maps to `$format` in the template.
    pub format: crate::types::AlbumFormat,
    /// Maps to `$recording` in the template.
    pub recording: crate::types::RecordingInfo,
}

/// Result of the statement parameterised by [`Input`].
pub type Output = OutputRow;

/// Row of [`Output`].
#[derive(Debug, Clone, PartialEq)]
pub struct OutputRow {
    /// Maps to the `id` result set column.
    pub id: i64,
}


impl crate::mapping::Statement for Input {
    type Result = Output;

    const RETURNS_ROWS: bool = true;

    const SQL: &str = "insert into album (name, released, format, recording)\n\
                       values ($1, $2, $3::public.album_format, $4::public.recording_info)\n\
                       returning id";

    const PARAM_TYPES: &'static [tokio_postgres::types::Type] = &[Type::TEXT, Type::DATE, Type::UNKNOWN, Type::UNKNOWN];

    #[allow(refining_impl_trait)]
    fn encode_params(
        &self,
    ) -> [&(dyn tokio_postgres::types::ToSql + Sync); Self::PARAM_TYPES.len()] {
        [&self.name, &self.released, &self.format, &self.recording]
    }

    fn decode_result(
        rows: Vec<tokio_postgres::Row>,
        _affected_rows: u64,
    ) -> Result<Self::Result, crate::mapping::DecodingError> {
        match rows.len() {
            0 => Err(crate::mapping::DecodingError::UnexpectedAmountOfRows {
                expected: 1,
                actual: 0,
            }),
            1 => {
                let row = rows.first().unwrap();
                Ok(OutputRow {
                    id: crate::mapping::decode_cell(row, 0, 0)?,
                })
            }
            n => Err(crate::mapping::DecodingError::UnexpectedAmountOfRows {
                expected: 1,
                actual: n,
            }),
        }
    }
}
